src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -383,17 +387,102 @@ instance [is_domain R] [normalized_gcd_monoid R] : is_domain W.coordinate_ring :
 
+variables (x : R) (y : R[X])
+
+/-- The class of the element $X - x$ in $R[W]$ for some $x \in R$. -/
+@[simp] noncomputable def X_class : W.coordinate_ring := adjoin_root.mk W.polynomial $ C $ X - C x
+
+/-- The class of the element $Y - y(X)$ in $R[W]$ for some $y(X) \in R[X]$. -/
+@[simp] noncomputable def Y_class : W.coordinate_ring := adjoin_root.mk W.polynomial $ X - C y
+
+/-- The ideal $\langle X - x \rangle$ of $R[W]$ for some $x \in R$. -/
+@[simp] noncomputable def X_ideal : ideal W.coordinate_ring := ideal.span {W.X_class x}
+
+/-- The ideal $\langle Y - y(X) \rangle$ of $R[W]$ for some $y(X) \in R[X]$. -/
+@[simp] noncomputable def Y_ideal : ideal W.coordinate_ring := ideal.span {W.Y_class y}
+
 end polynomial
 
 end weierstrass_curve
 
+namespace weierstrass_curve
+
+open polynomial
+
+open_locale non_zero_divisors polynomial
+
+variables {F : Type u} [field F] (W : weierstrass_curve F) (x : F) (y : F[X])
+
+instance coordinate_ring.is_domain_of_field : is_domain W.coordinate_ring :=
+by { classical, apply_instance }
+
+lemma X_class_ne_zero : W.X_class x ≠ 0 :=
+begin
+  intro hx,
+  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hx) with _ hx,
+  apply_fun degree at hx,
+  rw [degree_mul, polynomial_degree, degree_C $ X_sub_C_ne_zero x] at hx,
+  exact two_ne_zero (nat.with_bot.add_eq_zero_iff.mp hx).right
+end
+
+/-- The non-zero class of the element $X - x$ in $F(W)$ for some $x \in F$. -/
+@[simps] noncomputable def X_class' : W.function_fieldˣ :=
+units.mk0 _ $ (map_ne_zero_iff _ $ by exact no_zero_smul_divisors.algebra_map_injective _ _).mpr $
+  W.X_class_ne_zero x
+
+lemma Y_class_ne_zero : W.Y_class y ≠ 0 :=
+begin
+  intro hy,
+  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hy) with _ hy,
+  apply_fun degree at hy,
+  rw [degree_mul, polynomial_degree, degree_X_sub_C, nat.with_bot.add_eq_one_iff] at hy,
+  cases hy with hy hy,
+  { exact @order.succ_ne_succ (with_bot ℕ) _ _ _ _ _ one_ne_zero hy.right },
+  { exact two_ne_zero hy.right }
+end
+
+/-- The non-zero class of the element $Y - y(X)$ in $F(W)$ for some $y(X) \in F[X]$. -/
+@[simps] noncomputable def Y_class' : W.function_fieldˣ :=
+units.mk0 _ $ (map_ne_zero_iff _ $ by exact no_zero_smul_divisors.algebra_map_injective _ _).mpr $
+  W.Y_class_ne_zero y
+
+lemma X_ideal_mul_inv :
+  (W.X_ideal x : fractional_ideal W.coordinate_ring⁰ W.function_field) * (W.X_ideal x)⁻¹ = 1 :=
+fractional_ideal.coe_ideal_span_singleton_mul_inv W.function_field $ W.X_class_ne_zero x
+
+lemma X_ideal_inv_mul :
+  (W.X_ideal x : fractional_ideal W.coordinate_ring⁰ W.function_field)⁻¹ * W.X_ideal x = 1 :=
+fractional_ideal.coe_ideal_span_singleton_inv_mul W.function_field $ W.X_class_ne_zero x
+
+/-- The non-zero fractional ideal $\langle X - x \rangle$ of $F(W)$ for some $x \in F$. -/
+@[simps] noncomputable def X_ideal' : (fractional_ideal W.coordinate_ring⁰ W.function_field)ˣ :=
```

Comment:
I'm not convinced that you need to define `X_ideal'` and `Y_ideal'` as fractional ideals. I think in your #18034 you could prove that, if `x` and `y` are obtained by from `x' y' : ideal R` (`(x : fraction_ring ...) = x'`), then `class_group.mk x = class_group.mk y ↔ ∃ z w : R, z ≠ 0 ∧ w ≠ 0 ∧ z • x' = w • y'`, which would allow you to always work in `W.coordinate_ring`. It might be possible to always work in `F[X][X]` as well; you just need to adjoin `W.polynomial` into the ideals when considering equality between ideals, but that may be less convenient in some proofs.

Reply:
I haven't made sure everything still works but I'm now also convinced that we don't need the `_ideal'` definitions.